tests/integration/relations/opensearch_provider/application-charm/src/charm.py

```diff
@@ -0,0 +1,216 @@
+#!/usr/bin/env python3
+# Copyright 2023 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+"""Application charm that connects to opensearch using the opensearch-client relation."""
+
+import json
+import logging
+from typing import Dict, List, Optional, Union
+
+import requests
+from charms.data_platform_libs.v0.data_interfaces import (
+    DatabaseCreatedEvent,
+    DatabaseEndpointsChangedEvent,
+    DatabaseRequires,
+)
+from ops.charm import ActionEvent, CharmBase
+from ops.main import main
+from ops.model import ActiveStatus, BlockedStatus
+
+logger = logging.getLogger(__name__)
+
+
+class ApplicationCharm(CharmBase):
+    """Application charm that connects to database charms.
+
+    Enters BlockedStatus if it cannot constantly reach the database.
+    """
+
+    def __init__(self, *args):
+        super().__init__(*args)
+
+        # Default charm events.
+        self.framework.observe(self.on.update_status, self._on_update_status)
+
+        # Events related to the first database that is requested
+        # (these events are defined in the database requires charm library).
+        database_name = f'{self.app.name.replace("-", "_")}_first_database'
+
+        permissive_roles = json.dumps({"roles": ["all_access"]})
+        self.first_database = DatabaseRequires(
+            self, "first-database", database_name, permissive_roles
+        )
+        self.framework.observe(
+            self.first_database.on.database_created, self._on_first_database_created
+        )
+        self.framework.observe(
+            self.first_database.on.endpoints_changed, self._on_first_database_endpoints_changed
+        )
+
+        # Events related to the second database that is requested
+        # (these events are defined in the database requires charm library).
+        database_name = f'{self.app.name.replace("-", "_")}_second_database'
+        # TODO change this to include only permissions and action groups, and verify that we can
+        # create roles when necessary.
+        roles = {
+            "roles": ["readall"],
+            "permissions": ["TODO find some permissions", ""],
```

Comment:
Would be interesting to have different permissions for 2 different indices, and validate the permissions accordingly.

Reply:
This is the plan re: testing access control :)